### PR TITLE
Add Mean Field Variational Inference implementation 

### PR DESCRIPTION
**Pull request by xidulu**
Some initial attempts in integrating MFVI into blackjax https://github.com/blackjax-devs/blackjax/issues/397.

TODO:
There is going to be lots of boiler plate code shared between MFVI and Fullrank VI.

It is also worth considering how to add stick-the-landing gradient estimator [1] and importance-weighted ELBO optimization [2].

[1] Roeder, Geoffrey, Yuhuai Wu, and David K. Duvenaud. "Sticking the landing: Simple, lower-variance gradient estimators for variational inference." Advances in Neural Information Processing Systems 30 (2017).

[2] Domke, Justin, and Daniel R. Sheldon. "Importance weighting and variational inference." Advances in neural information processing systems 31 (2018).

### CODE REVIEW

File: blackjax/base.py [hunk 1]
```diff
@@ -133,11 +133,15 @@ class VIAlgorithm(NamedTuple):
     sample
         A pure function which returns samples from the approximation computed
         by `approximate`.
+    log_prob
```
Comment (junpenglao) [inline blackjax/base.py:136]:
```suggestion
    logdensity_fn
```
(Same elsewhere)

File: blackjax/vi/mfvi.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    def meanfield_logprob(position):
+        logq_pytree = jax.tree_map(stats.norm.logpdf, position, mu_param, sigma_param)
+        logq = jax.tree_map(jnp.sum, logq_pytree)
+        return jax.tree_util.tree_reduce(jnp.add, logq)
+
+    return meanfield_logprob
+
+
+def sample(rng_key, meanfield_param, num_samples: int):
```
Comment (junpenglao) [inline blackjax/vi/mfvi.py:45]: Maybe you can call https://github.com/blackjax-devs/blackjax/blob/d6801f8a74721f881b5699bd536eea863e2a0009/blackjax/util.py#L56 internally?

Comment (xidulu) [reply blackjax/vi/mfvi.py:45]: Sounds good.
But at this moment there's no internal function we can call to sample from multivariate Gaussian right (i.e. the fullrank VI case)?

**Review by rlouf — COMMENTED** [review]
Thanks for contributing! That's a very good start, we'll just need to reorganise a few things.

File: blackjax/vi/mfvi.py (hunk 2 above)
Comment (junpenglao) [reply blackjax/vi/mfvi.py:45]: You can use the same function for multivariate Gaussian because the `linear_map` util will dispatch it correctly.

File: tests/test_mfvi.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import functools
+import chex
+import jax
+import jax.numpy as jnp
+import jax.scipy.stats as stats
+import numpy as np
+import blackjax
+import optax
```
[the remark this answers is not in the document]
Comment (junpenglao) [reply tests/test_mfvi.py:8]: I thought `optax` is used in SGMCMC? I think it is ok to add it as dependency.

File: blackjax/kernels.py [hunk 4]
```diff
@@ -1335,3 +1335,26 @@ def kernel(rng_key, state):
         return AdaptationResults(last_chain_state, kernel, parameters)
 
     return AdaptationAlgorithm(run)
+
```
[the remark this answers is not in the document]
Comment (xidulu) [reply blackjax/kernels.py:1338]: My motivation is that the users could possibly need to use the log_q function for further downstream operations such as importance-weighted sampling or simply density estimation (let's say we also implement normalization flow one day). Also, one of the biggest advantages of VI is that it has a parametric approximate posterior with tractable density. If we don't return it to users, then the users would only have access to a black box sampler.

File: blackjax/vi/mfvi.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            logp = vmap(log_prob_fn)(z)
+            return (logq - logp).mean()
+
+        # compute KL divergence
+        elbo, grad = jax.value_and_grad(kl_fn)(params)
+        updates, opt_state = optimizer.update(grad, opt_state, params)
+        params = jax.tree_map(lambda p, u: p + u, params, updates)
+        return (params, opt_state), elbo
+
+    def run_optimization(init_params):
```
[the remark this answers is not in the document]
Comment (xidulu) [reply blackjax/vi/mfvi.py:79]: Sounds good

File: blackjax/vi/mfvi.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        jax.random.normal(rng_key, num_samples + mu_flatten.shape) * sigma_flatten
+        + mu_flatten
+    )
+    if len(num_samples) == 0:
+        return unravel_fn(flatten_sample)
+    return jax.vmap(unravel_fn)(flatten_sample)
+
+
+def approximate(
+    rng, init_params, log_prob_fn, optimizer, sample_size=5, num_steps=200
```
[the remark this answers is not in the document]
Comment (xidulu) [reply blackjax/vi/mfvi.py:62]: The usual practice is to use Adam which is available in both Optax and Jaxopt. I will have a hold here till you and @junpenglao decide which optimization library to use for VI.

File: tests/test_mfvi.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+            (2, 4),
+            (3, 5)
+        ]
+        def logdensity_fn(x):
+            logpdf = (stats.norm.logpdf(x["x_1"], *ground_truth[0]) +
+                        stats.norm.logpdf(x["x_2"], *ground_truth[1]))
+            return jnp.sum(logpdf)
+
+        initial_position = {"x_1": 0.0, "x_2": 0.0}
+        mu_param = jax.tree_map(jnp.zeros_like, initial_position)
```
[the remark this answers is not in the document]
Comment (xidulu) [reply tests/test_mfvi.py:29]: Agreed with that, I propose to add a static method like this:

```python
    @staticmethod
    def init_params_fn(initial_position: PyTree) -> Tuple[PyTree, PyTree]:
        mu_param = jax.tree_map(jnp.zeros_like, initial_position)
        # Initial the scale parameter with a small value.
        rho_param = jax.tree_map(-2.0 * jnp.ones_like, initial_position)
        return (mu_param, rho_param)
```

which takes in a PyTree that stores the latent variables and returns an initial value of the variational parameters.

And then we could use it like:

```python
initial_position = {"x_1": 0.0, "x_2": 0.0}
kernel = blackjax.kernels.meanfield_vi(logdensity_fn)
 result = kernel.approximate(
            key,
            blackjax.kernels.meanfield_vi.init_params_fn(initial_position),
            optimizer, num_steps=50000, sample_size=500
        )
```

File: blackjax/kernels.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rlouf) [reply blackjax/kernels.py:1338]: This all makes sense. Since this is an advanced use case, people could import the function from `blackjax.vi.mfdvi`

[the remark this answers is not in the document]
Comment (xidulu) [reply blackjax/kernels.py:1338]: That sounds reasonable!

File: blackjax/vi/mfvi.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+from jax.scipy import stats
+from jax import vmap
+
+
+__all__ = ["approximate", "sample", "mfvi_logprob", "MFVIState"]
+
+
+class MFVIState(NamedTuple):
+    elbo: Array
+    position: PyTree
```
[the remark this answers is not in the document]
Comment (xidulu) [reply blackjax/vi/mfvi.py:30]: Yes position corresponds to mu and rho, I will change it.

File: blackjax/vi/mfvi.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+import jax.random
+from jax.scipy import stats
+from jax import vmap
+
+
+__all__ = ["approximate", "sample", "mfvi_logprob", "MFVIState"]
+
+
+class MFVIState(NamedTuple):
+    elbo: Array
```
[the remark this answers is not in the document]
Comment (xidulu) [reply blackjax/vi/mfvi.py:29]: Not really, but people would need the ELBO trace to inspect the convergence of the optimiaztion

[the remark this answers is not in the document]
Comment (rlouf) [reply blackjax/vi/mfvi.py:29]: It should be returned as info then. The pattern we're trying to follow is return in the state everything that's needed for further computations, and in an info tuple additional information that can be used e.g. for inspecting.

Here the state would be what's needed to draw samples, the current state of `mu` and `rho`, and the information could be all the values of `mu`, `rho` and the `elbo` seen during the optimisation phase. What do you think?

[the remark this answers is not in the document]
Comment (xidulu) [reply blackjax/vi/mfvi.py:29]: > t should be returned as info then. The pattern we're trying to follow is return in the state everything that's needed for further computations, and in an info tuple additional information that can be used e.g. for inspecting.

Sounds good!

**Comment by rlouf** [comment]
I think that's enough comments for now :) Don't be discouraged, this is completely normal as we are also figuring out what the interface should be; there is no template as there can be for MCMC algorithms yet.

File: blackjax/vi/mfvi.py (hunk 2 above)
Comment (xidulu) [reply blackjax/vi/mfvi.py]: @junpenglao But how do you generate num_sample of particles using `generate_gaussian_noise`

**Comment by xidulu** [comment]
@rlouf @junpenglao Would you mind taking another look at the API?

I think the next thing we need to figure out is to avoid potential boilerplate code since the ELBO computation and optimization process is identical for most of the VI variants.

File: blackjax/vi/mfvi.py (hunk 2 above)
Comment (junpenglao) [reply blackjax/vi/mfvi.py]: You are right the util function does not accept that yet - let's keep your version here but could you add a TODO?
```
# TODO: switch to using `generate_gaussian_noise` in util.py
```

**Comment by rlouf** [comment]
> I think the next thing we need to figure out is to avoid potential boilerplate code since the ELBO computation and optimization process is identical for most of the VI variants. 

I prefer to avoid generalising on the first implementation even though it indeed seems like we would need the equivalent of `base.py` with the SMC algorithm.

I'm taking some time off and won't be able to take a good look at the PR until Jan 3, but will have plenty of time that day to go over it in depth.

**Comment by xidulu** [comment]
@rlouf Well, apparently I need better documentation and more detailed test cases. Other than that, do you have any further comments on the API design? 

Also, I am thinking about moving forward to implementing the full-rank VI, do you think I should open up a new PR, or should I just implement it in this PR.

**Comment by rlouf** [comment]
I am making a pass on your implementation atm; you'll need to pull the changes once I am finished (`pull --rebase`), but I'm not done quite yet.

I now think your original proposal of implementing VI as a "kernel" was the right way to go, and we will probably need to "kernelize" (we need a better name) Pathfinder as well. I think it's nice that the user has fine-grained control over the optimization.

**Comment by rlouf** [comment]
I rearranged things a little, it was mostly cosmetic. I think the next step is to give it a kernel-like API, tests, docs. If the full-rank is going to share a lot of code it's probably best to do it now rather than go through the hassle of opening a new PR.

**Comment by rlouf** [comment]
I turned the algorithm into a "kernel" like you originally suggested, I think it is much more in the general spirit of Blackjax and gives the user more freedom. This is almost ready to merge, we need to expand the docstring of the step function a little more, and add references.

**Comment by xidulu** [comment]
@rlouf Thanks! I can take care of some documentation jobs! See you TMR in the meeting!

File: blackjax/base.py [hunk 10]
```diff
@@ -1,5 +1,5 @@
 # Copyright 2020- The Blackjax Authors.
-#
+# , Tuple
```
Comment (junpenglao) [inline blackjax/base.py:2]: Remove

File: blackjax/vi/meanfield_vi.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    mu_flatten, unravel_fn = jax.flatten_util.ravel_pytree(mu)
+    sigma_flat, _ = jax.flatten_util.ravel_pytree(sigma)
+    flatten_sample = (
+        jax.random.normal(rng_key, (num_samples,) + mu_flatten.shape) * sigma_flat
+        + mu_flatten
+    )
+    return jax.vmap(unravel_fn)(flatten_sample)
+
+
+def logprob(mu, rho):
```
Comment (junpenglao) [inline blackjax/vi/meanfield_vi.py:113]:
```suggestion
def generate_meanfield_logprob(mu, rho):
```

**Comment by rlouf** [comment]
This looks great! I think it is ready to merge after we replace `logprob` with `logdensity` 😊

**Comment by xidulu** [comment]
@rlouf I got it. One second

**Comment by xidulu** [comment]
@rlouf I just made some changes, is that what you suggest?

**Comment by rlouf** [comment]
I meant `logdensity` instead of `logdensity_fn` sorry. Then it'd be great if you could squash your commits. Let me known, and we'll merge if the tests pass... and move on to full rank :)

**Comment by xidulu** [comment]
@rlouf Done

**Comment by rlouf** [comment]
LGTM, thank you for contributing!
